---
title: "Global Coral Microbiome Project: Australia, Hawaii, Colombia and Panama"
authors: "Colin Howe"
date: "10/05/2023"
output: html_document
---

# First, Set working Directory load library and Import Data into Phyloseq
```{r warning=FALSE}
# Command for setting your work directory
setwd("C:/Users/Colin Howe/Desktop/PSU Doctoral Program/GCMP_Working/gcmp_qiime_input")

# Print your working directory

getwd()
#
```

## Load required libraries. 
```{r, message=FALSE, warning=FALSE} 
library(BiocManager)
library(phyloseq)
library(dplyr)
library(qiime2R)
library(DESeq2)
library(ComplexHeatmap)
library(ade4)
library(vegan)
library(RColorBrewer)
library(tidyverse)
library(ggplot2)
library(microbiome)
library(dendextend)
library(gplots)
library(ggrepel)
library(tidyr)
library(ggpubr)
library(cowplot)
library(RVAideMemoire)
library(ape)
library(microbiomeMarker)
```

# Let's cleaning the taxa file!
```{r}
##Creating require files for phyloseq
#### Create ASV and taxonomy from qza tables importing as a matrix
##This was created using the biom export function to create a features.tsv file

#simple import from qza to phyloseq
asv <- qza_to_phyloseq(features = "gcmp_qiime_input/filtered_merged_table_v2.qza")

#### Import Metadata read.table
metadata <- read.table(file = "gcmp_qiime_input/geo_bon_mapping.csv",header=T,comment.char="", row.names=1, sep=",")

### Import Tree file from biom output tree.nwk
tree2 <- read_tree("gcmp_qiime_input/tree.nwk")

### Import taxonomy * read.table
#taxonomy2 <- qza_to_phyloseq(taxonomy = "qiime_results/rush_taxonomy.qza")
taxonomy <- read.table(file = "gcmp_qiime_input/taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

# clean the taxonomy, Greengenes format
tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "k__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}


OTU <- otu_table(as.matrix(asv), taxa_are_rows = TRUE)
tax1 = tax_table(as.matrix(tax.clean))
SAMPLE <- sample_data(metadata)
TREE = read_tree(tree2)

ch_phylo <- phyloseq(OTU,tax1,SAMPLE,TREE)


#nsamples(ch_phylo)
#sample_data(ch_phylo)
#sample_variables(ch_phylo)
#sample_names(ch_phylo)

```

## Clean up not target sequences the dataset 
```{r, message=FALSE, warning=FALSE}
library(gridExtra)
# Remove features that have only been classified at the kingdom level
phylo_noKing <-subset_taxa(ch_phylo, kingdom != "d__Eukaryota")

phylo_noNA <- subset_taxa(phylo_noKing,Phylum != "NA", Order!= "Chloroplast" & Family != "Mitochondria" & Order !="Unclassified bacteroidia")

# Remove any samples with less than 1 read
phylo_coral = prune_samples(sample_sums(phylo_noNA)>=1, phylo_noNA)
phylo_coral

## After we now have 583 unique bacterial and archaea families
#phylo_nclass = subset_taxa(phylo_nclass, Family !="Unclassified Proteobacteria")

##Rarefication
set.seed(111) # keep result reproductive
gcmp.rarefied = rarefy_even_depth(phylo_coral, rngseed=1, sample.size=1000, replace=T) ## After looking up rarefy_ command the notes advise against using this command. Instead they suggest phyloseq_to_deseq2 command for linear mixed effects models. 
gcmp.rarefied
#get_taxa_unique(phylo_nlow, "Family")

# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(phylo_coral),
               MARGIN = ifelse(taxa_are_rows(phylo_coral), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
#Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(phylo_coral),
                    tax_table(phylo_coral))

#Compute the total and average prevalences of the features in each phylum(Family)
coral_family_prev <-plyr::ddply(prevdf, "Family", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

## After looking at the dataframe I noticed there are several ASV that have 0 prevalence and 0 abundance. I filtered these out as well. 0.05 represents 5% of the dataset. However this contains on 13 families. Instead I used 0.5% and it returned 202 Asv families. Additionally any prevalence under 0.05 filters a majority of taxa found. from 17k to 1k
#phylo_prev = 0.005* nsamples(phylo_coral)
#keepTaxa = rownames(prevdf)[(prevdf$Prevalence >= phylo_prev)]
#phylo_nlow_prev = prune_taxa(keepTaxa, phylo_coral)

## Check Results
#get_taxa_unique(phylo_nlow1, "Family")
#get_taxa_unique(phylo_nlow_prev, "Family")

#multiPlotTitleTextSize = 15
#p2tree = plot_tree(phylo_nlow, method = "treeonly",
#                   ladderize = "left",
#                   title = "Taxa Count = 1") +
#  theme(plot.title = element_text(size = multiPlotTitleTextSize))
#p3tree = plot_tree(phylo_nclass, method = "treeonly",
#                   ladderize = "left", title = "Unclassified Bacteria \n & Archaea") +
#  theme(plot.title = element_text(size = multiPlotTitleTextSize))
#p4tree = plot_tree(phylo_nlow_prev, method = "treeonly",
#                   ladderize = "left", title = "Minimum Prevelence \n of 0.1%") +
#  theme(plot.title = element_text(size = multiPlotTitleTextSize))
#p5tree = plot_tree(ps.rarefied, method = "treeonly",
#                   ladderize = "left", title = "Rarefied to\n 2000 taxa") +
#  theme(plot.title = element_text(size = multiPlotTitleTextSize))
#
#grid.arrange(nrow = 1, p2tree, p3tree, p4tree, p5tree)


#phylo_nlow
#phylo_nclass
#phylo_nlow_prev
#ps.rarefied
## Rarefied dataset contains 553 families compared to 583. (30 families lost) 
#get_taxa_unique(ps.rarefied, "Family")


```

### Subset Phyloseq Objects for Comparative Analysis 
```{r}
## So far aus complex group show trends however fussy with acropora genus so I subset that and Pachyseris
## Aus acropora genus tissue and skeleton across bray-wuni show one taxa explains most variation due to sample size

### Subsets with All samples with some that exclude out-groups and expeditions
main_coral <- subset_samples(gcmp.rarefied, Biological_group == "yes" & country_group == "yes" & coral_group =="yes") # All with outgroups millepora and stylerasteridae 

##Sub_sample Corals with no outgroup
coral <-subset_samples(main_coral, coral_group == "yes") #corals without outgroups and environmental samples 
## Regional Intra vs Inter Specific Variation between species, genus, family & Clade

####### Australia ######
## Separation by Location and Clade and compartment
#  Regional Within Clade variation
complex_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & complex_robust == "complex")#n=42
complex_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & complex_robust == "complex") #n=49

#  Regional Within Family variation
acroporidae_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & family == "Acroporidae")#n=14
acroporidae_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Acroporidae")#n=20

#  Regional Within Genus variation
acropora_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & host_genus_id == "Acropora")#n=12
acropora_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & host_genus_id == "Acropora")#n=17

#  Regional Within Genus variation
Pachyseris_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & host_genus_id == "Pachyseris")#n=8
Pachyseris_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & host_genus_id == "Pachyseris")#n=6


## Robust Comparision
# Regional Within Clade variation
robust_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & complex_robust == "robust")#n=34
robust_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & complex_robust == "robust")#n=41

#  Regional Within Family variation
merulinidae_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & family == "Merulinidae")#n=7
merulinidae_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Merulinidae")#n=8

#  Regional Within Family variation
pocilloporidae_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & family == "Pocilloporidae")#n=3
# only has one representative pocilloporidae_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Pocilloporidae")#n=1

#  Regional Within Genus variation
Echinopora_aus_tis <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Tissue" & host_genus_id == "Echinopora")#n=3
Echinopora_aus_ske <-subset_samples(main_coral, country_region == "Australia_Pacific" & BiologicalMatter == "Coral Skeleton" & host_genus_id == "Echinopora")#n=3


####### Caribbean ######
## Separation by Location and Clade and compartment
#  Regional Within Clade variation
complex_carib_tis <-subset_samples(main_coral, ocean_area == "Caribbean" & BiologicalMatter == "Coral Tissue" & complex_robust == "complex")#n=11
## No samples passed rareification  complex_carib_ske <-subset_samples(main_coral, ocean_area == "Caribbean" & BiologicalMatter == "Coral Skeleton") & complex_robust == "complex")

#  Regional Within Family variation
acroporidae_cairb_tis <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Tissue" & family == "Acroporidae")#n=10
## no samples passed rareification acroporidae_cairb_ske <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Skeleton" & family == "Acroporidae")

#  Regional Within Family variation
poritidae_aus_tis <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Tissue" & family == "Poritidae")#n=4
poritidae_aus_ske <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Skeleton" & family == "Poritidae")#n=4 

## Robust Comparision
# Regional Within Clade variation Orbicella represent robust clade from this comparison no merulinidae comparison needed
robust_carib_tis <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Tissue" & complex_robust == "robust")#n=30

robust_carib_ske <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Skeleton" & complex_robust == "robust")#n=22

#  Regional Within Genus variation ## I need to subset into just Colombia because of regional effect
orbicella_tis <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Tissue" & host_genus_id == "Orbicella")#n=11
orbicella_ske <-subset_samples(main_coral,  ocean_area == "Caribbean" & BiologicalMatter == "Coral Skeleton" & host_genus_id == "Orbicella")#n=11


####### Hawaii ######
## Separation by Location and Clade and compartment
#  Regional Within Clade variation
complex_hawaii_tis <-subset_samples(main_coral, ocean_area == "North Pacific" & BiologicalMatter == "Coral Tissue" & complex_robust == "complex")#n=10
complex_hawaii_ske <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Skeleton" & complex_robust == "complex")#n=11

#  Regional Within Family variation
agariciidae_hawaii_tis <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Tissue" & family == "Agariciidae")#n=3
agariciidae_hawaii_ske <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Agariciidae")#n=3

#  Regional Within Family variation
poritidae_hawaii_tis <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Tissue" & family == "Poritidae")#n=4
poritidae_hawaii_ske <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Poritidae")#n=4


## Robust Comparision
# Regional Within Clade variation
robust_hawaii_tis <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Tissue" & complex_robust == "robust")#n=7
robust_hawaii_ske <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Skeleton" & complex_robust == "robust")#n=8

#  Regional Within Family variation
pocilloporidae_hawaii_tis <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Tissue" & family == "Pocilloporidae")#n=5
pocilloporidae_hawaii_ske <-subset_samples(main_coral,  ocean_area == "North Pacific" & BiologicalMatter == "Coral Skeleton" & family == "Pocilloporidae")#n=4

#  Across Panova Genus variation
pavona_tis <-subset_samples(main_coral, BiologicalMatter == "Coral Tissue" & host_genus_id == "Pavona")#n=17
pavona_ske <-subset_samples(main_coral, BiologicalMatter == "Coral Skeleton" & host_genus_id == "Pavona")#n=20

```

## Neutral Model Caribbean
```{r}
####NCM Neutral model############################

#rm(list=ls())#clear previous data
#Adam Burns - 2/10/2015
#aburns2@uoregon.edu

#From Burns et al. Contribution of neutral processes to the assembly of the gut microbial communities changes over host development

#Fits the neutral model from Sloan et al. 2006 to an OTU table and returns several fitting statistics. Alternatively, will return predicted occurrence frequencies for each OTU based on their abundance in the metacommunity when stats=FALSE. For use in R.

#spp: A community table for communities of interest with local communities/samples as rows and taxa as columns. All samples must be rarefied to the same depth
#coral <- subset_samples(gcmp.rarefied, Biological_group == "yes" & coral_group =="yes") # All with outgroups millepora and stylerasteridae 


##Sub_sample Corals Acropora
acro <-subset_samples(ch_phylo, ocean_area == "Caribbean" & family == "Acroporidae")
acro
set.seed(111) # keep result reproductive
acro.rarefied = rarefy_even_depth(acro, rngseed=1, sample.size=3000, replace=T)
acro.rarefied
acro

## Subsample Meru
meru <-subset_samples(ch_phylo, ocean_area == "Caribbean" & family == "Merulinidae")
set.seed(111) # keep result reproductive
meru.rarefied = rarefy_even_depth(meru, rngseed=1, sample.size=3000, replace=T)
meru.rarefied
meru
## Environmental Samples 
env_sample <- subset_samples(ch_phylo, Biological_group == "no" & ocean == "Atlantic")
env_sample

set.seed(111) # keep result reproductive
env.rarefied = rarefy_even_depth(env_sample, rngseed=1, sample.size=3000, replace=T)
env.rarefied
env_sample

##### Subsample complex clade
complex <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "complex")

set.seed(111)
complex.rarefied = rarefy_even_depth(complex, rngseed=1, sample.size=3000, replace=T)
complex.rarefied
complex

## Subsample complex mucus
complex.m <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "complex" & BiologicalMatter == "Coral Mucus")

set.seed(111)
complex.m.rarefied = rarefy_even_depth(complex.m, rngseed=1, sample.size=3000, replace=T)
complex.m.rarefied
complex.m

## Subsample complex tissue
complex.s <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "complex"& BiologicalMatter == "Coral Skeleton")

set.seed(111)
complex.s.rarefied = rarefy_even_depth(complex.s, rngseed=1, sample.size=3000, replace=T)
complex.s.rarefied
complex.s


##### Subsample Robust clade
robust <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "complex")

set.seed(111)
robust.rarefied = rarefy_even_depth(robust, rngseed=1, sample.size=3000, replace=T)
robust.rarefied
robust

## Subsample robust
robust.m <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "robust" & BiologicalMatter == "Coral Mucus" )

set.seed(111)
robust.m.rarefied = rarefy_even_depth(robust.m, rngseed=1, sample.size=3000, replace=T)
robust.m.rarefied


## Subsample robust
robust.s <-subset_samples(ch_phylo, ocean_area == "Caribbean" & complex_robust == "robust"& BiologicalMatter == "Coral Skeleton")

set.seed(111)
robust.s.rarefied = rarefy_even_depth(robust.s, rngseed=1, sample.size=3000, replace=T)
robust.s.rarefied
robust.s
#pool: A community table for defining source community (optional; Default=NULL).

#taxon: A table listing the taxonomic calls for each otu, with OTU ids as row names and taxonomic classifications as columns.

#If stats=TRUE the function will return fitting statistics.
#If stats=FALSE the function will return a table of observed and predicted values for each otu.

## Plot distribution
glom_meru <- tax_glom(meru.rarefied, taxrank = 'Family', NArm = TRUE)
glom_acro <- tax_glom(acro.rarefied, taxrank = 'Family', NArm = TRUE)
glom_env <- tax_glom(env.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp <- tax_glom(complex.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob <- tax_glom(robust.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp_m <- tax_glom(complex.m.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp_s <- tax_glom(complex.s.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob_m <- tax_glom(robust.m.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob_s <- tax_glom(robust.s.rarefied, taxrank = 'Family', NArm = TRUE)


################define sncm.fit############
sncm.fit <- function(spp, pool=NULL, stats=TRUE, taxon=NULL){
	require(minpack.lm)
	require(Hmisc)
	require(stats4)
	
	options(warn=-1)

	#Calculate the number of individuals per community
	N <- mean(apply(spp, 1, sum))
	
	#Calculate the average relative abundance of each taxa across communities
	if(is.null(pool)){
		p.m <- apply(spp, 2, mean)
		p.m <- p.m[p.m != 0]
		p <- p.m/N
	} else {
		p.m <- apply(pool, 2, mean)
		p.m <- p.m[p.m != 0]
		p <- p.m/N
	}

#Calculate the occurrence frequency of each taxa across communities
spp.bi <- 1*(spp>0)
freq <- apply(spp.bi, 2, mean)
freq <- freq[freq != 0]
#Combine
C <- merge(p, freq, by=0)
C <- C[order(C[,2]),]
C <- as.data.frame(C)
C.0 <- C[!(apply(C, 1, function(y) any(y == 0))),] #Removes rows with any zero (absent in either source pool or local communities)
p <- C.0[,2]
freq <- C.0[,3]
names(p) <- C.0[,1]
names(freq) <- C.0[,1]

#Calculate the limit of detection
d = 1/N

##Fit model parameter m (or Nm) using Non-linear least squares (NLS)
m.fit <- nlsLM(freq ~ pbeta(d, N*m*p, N*m*(1-p), lower.tail=FALSE), start=list(m=0.1))
m.ci <- confint(m.fit, 'm', level=0.95)
	
##Fit neutral model parameter m (or Nm) using Maximum likelihood estimation (MLE)
sncm.LL <- function(m, sigma){
	R = freq - pbeta(d, N*m*p, N*m*(1-p), lower.tail=FALSE)
	R = dnorm(R, 0, sigma)
	-sum(log(R))
}
	
m.mle <- mle(sncm.LL, start=list(m=0.1, sigma=0.1), nobs=length(p))

##Calculate Akaike's Information Criterion (AIC)
aic.fit <- AIC(m.mle, k=2)
bic.fit <- BIC(m.mle)

##Calculate goodness-of-fit (R-squared and Root Mean Squared Error)
freq.pred <- pbeta(d, N*coef(m.fit)*p, N*coef(m.fit)*(1-p), lower.tail=FALSE)
Rsqr <- 1 - (sum((freq - freq.pred)^2))/(sum((freq - mean(freq))^2))
RMSE <- sqrt(sum((freq-freq.pred)^2)/(length(freq)-1))
	
pred.ci <- binconf(freq.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)
	
##Calculate AIC for binomial model
bino.LL <- function(mu, sigma){
R = freq - pbinom(d, N, p, lower.tail=FALSE)
R = dnorm(R, mu, sigma)
-sum(log(R))
	}
bino.mle <- mle(bino.LL, start=list(mu=0, sigma=0.1), nobs=length(p))
	
aic.bino <- AIC(bino.mle, k=2)
bic.bino <- BIC(bino.mle)
	
##Goodness of fit for binomial model
bino.pred <- pbinom(d, N, p, lower.tail=FALSE)
Rsqr.bino <- 1 - (sum((freq - bino.pred)^2))/(sum((freq - mean(freq))^2))
RMSE.bino <- sqrt(sum((freq - bino.pred)^2)/(length(freq) - 1))

bino.pred.ci <- binconf(bino.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)
	
##Calculate AIC for Poisson model
pois.LL <- function(mu, sigma){
	R = freq - ppois(d, N*p, lower.tail=FALSE)
	R = dnorm(R, mu, sigma)
	-sum(log(R))
}
pois.mle <- mle(pois.LL, start=list(mu=0, sigma=0.1), nobs=length(p))
	
aic.pois <- AIC(pois.mle, k=2)
bic.pois <- BIC(pois.mle)
	
##Goodness of fit for Poisson model
pois.pred <- ppois(d, N*p, lower.tail=FALSE)
Rsqr.pois <- 1 - (sum((freq - pois.pred)^2))/(sum((freq - mean(freq))^2))
RMSE.pois <- sqrt(sum((freq - pois.pred)^2)/(length(freq) - 1))

pois.pred.ci <- binconf(pois.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)

##Results
	if(stats==TRUE){
		fitstats <- data.frame(m=numeric(), m.ci=numeric(), m.mle=numeric(), maxLL=numeric(), binoLL=numeric(), poisLL=numeric(), Rsqr=numeric(), Rsqr.bino=numeric(), Rsqr.pois=numeric(), RMSE=numeric(), RMSE.bino=numeric(), RMSE.pois=numeric(), AIC=numeric(), BIC=numeric(), AIC.bino=numeric(), BIC.bino=numeric(), AIC.pois=numeric(), BIC.pois=numeric(), N=numeric(), Samples=numeric(), Richness=numeric(), Detect=numeric())
		fitstats[1,] <- c(coef(m.fit), coef(m.fit)-m.ci[1], m.mle@coef['m'], m.mle@details$value, bino.mle@details$value, pois.mle@details$value, Rsqr, Rsqr.bino, Rsqr.pois, RMSE, RMSE.bino, RMSE.pois, aic.fit, bic.fit, aic.bino, bic.bino, aic.pois, bic.pois, N, nrow(spp), length(p), d)
		return(fitstats)
	} else {
	A <- cbind(p, freq, freq.pred, pred.ci[,2:3], bino.pred, bino.pred.ci[,2:3])
	A <- as.data.frame(A)
		colnames(A) <- c('p', 'freq', 'freq.pred', 'pred.lwr', 'pred.upr', 'bino.pred', 'bino.lwr', 'bino.upr')
		if(is.null(taxon)){
			B <- A[order(A[,1]),]
		} else {
			B <- merge(A, taxon, by=0, all=TRUE)
			row.names(B) <- B[,1]
			B <- B[,-1]
			B <- B[order(B[,1]),]
		}
		return(B)
	}
}
##################################

set.seed(111)
###coral communities
coral <- prune_taxa(taxa_sums(glom_comp) > 0, glom_comp)
coral
otu_coral = as.data.frame(t(otu_table(coral)))

mod_w_coral = sncm.fit(otu_coral)
mod1_w_coral = sncm.fit(otu_coral,stats = F)

##Identify above, below and neutral OTUs 
addf = function(freq, pred.upr, pred.lwr) {
  if(freq > pred.upr){
    model = "above"
  }
  else if(freq < pred.lwr){
    model = "below"
  }
  else if(pred.upr >= freq & freq >= pred.lwr){
    model = "neutral"
  }
  return(model)
}

mod1_w_coral$model = mapply(addf, mod1_w_coral$freq, mod1_w_coral$pred.upr, mod1_w_coral$pred.lwr)  
length(which(mod1_w_coral$model == "above"))
length(which(mod1_w_coral$model == "below"))
length(which(mod1_w_coral$model == "neutral"))

mod1_w_coral$model <- factor(mod1_w_coral$model, 
                             levels = c("above", "neutral", "below"))

## Assign Taxonomy to table of significant microbes
coral_taxa = cbind(as(mod1_w_coral, "data.frame"), as(tax_table(glom_comp)[rownames(mod1_w_coral), ], "matrix"))

set.seed(111)
###seawater communities
#seawater <- subset_samples(physeq2, SampleType == "seawater")
seawater <- prune_taxa(taxa_sums(glom_rob) > 0, glom_rob)
seawater

otu_seawater = as.data.frame(t(otu_table(seawater)))

mod_w_seawater = sncm.fit(otu_seawater)
mod1_w_seawater = sncm.fit(otu_seawater,stats = F)

##Identify above, below and netural OTUs 
addf = function(freq, pred.upr, pred.lwr) {
  if(freq > pred.upr){
    model = "above"
  }
  else if(freq < pred.lwr){
    model = "below"
  }
  else if(pred.upr >= freq & freq >= pred.lwr){
    model = "neutral"
  }
  return(model)
}

mod1_w_seawater$model = mapply(addf, mod1_w_seawater$freq, mod1_w_seawater$pred.upr, mod1_w_seawater$pred.lwr)  
length(which(mod1_w_seawater$model == "above"))
length(which(mod1_w_seawater$model == "below"))
length(which(mod1_w_seawater$model == "neutral"))

mod1_w_seawater$model <- factor(mod1_w_seawater$model, 
                                levels = c("above", "neutral", "below"))

## Assign Taxonomy to table of significant microbes
env_taxa = cbind(as(mod1_w_seawater, "data.frame"), as(tax_table(glom_rob)[rownames(mod1_w_seawater), ], "matrix"))


#plot neuModel
p_neuM_coral=ggplot(mod1_w_coral, aes(x=log10(p), y=freq, color = model))+
  geom_point(shape=19, alpha=0.7, size=2.5)+
  geom_line(aes(x=log10(p), y=freq.pred),
            color="blue",linetype="solid",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.lwr),
            color="blue",linetype="dashed",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.upr),
            color="blue",linetype="dashed",linewidth=1)+
  scale_colour_manual(name="OTUs",
                      values=c("#E41A1C", "#999999", "#FF7F00"),
                      breaks=c("above", "neutral", "below"),
                      labels=c("Over-represented", "Neutrally-distributed", "Under-represented"))+
  xlab("Log10 (Mean Relative abundance)") + ylab("Occurance frequency")+
  theme_bw() +
  theme(axis.title.x = element_text(face="bold", size=15),
        axis.title.y = element_text(face="bold", size=15),
        axis.text.x = element_text(colour = "black", size=12),
        axis.text.y = element_text(colour = "black", size=12), 
        legend.title = element_text(size=11.5, face = "bold"),
        legend.text = element_text(size=11.5),
        legend.position="none")

lb1 <- expression(paste(R^2,"0.3415789 Complex Clade"))
lb2 <- expression(paste(m," 0.007656227 Complex Clade"))
p_neuM_coral1 = p_neuM_coral + annotate("text", x = -3.5, y=0.98, size=6.5, label=lb1, 
                                        fontface="bold", color="black",
                                        parse=TRUE) +
  annotate("text", x = -3.5, y=0.9, size=6.5, label=lb2, 
           fontface="bold", color="black",
           parse=TRUE)


#plot neuModel
p_neuM_seawater=ggplot(mod1_w_seawater, aes(x=log10(p), y=freq, color = model))+
  geom_point(shape=19, alpha=0.7, size=2.5)+
  geom_line(aes(x=log10(p), y=freq.pred),
            color="blue",linetype="solid",size=1)+
  geom_line(aes(x=log10(p), y=pred.lwr),
            color="blue",linetype="dashed",size=1)+
  geom_line(aes(x=log10(p), y=pred.upr),
            color="blue",linetype="dashed",size=1)+
  scale_colour_manual(name="OTUs",
                      values=c("#E41A1C", "#999999", "#FF7F00"),
                      breaks=c("above", "neutral", "below"),
                      labels=c("Over-represented", "Neutrally-distributed", "Under-represented"))+
  xlab("Log10 (Mean Relative abundance)") + ylab("Occurance frequency")+
  theme_bw() +
  theme(axis.title.x = element_text(face="bold", size=15),
        axis.title.y = element_text(face="bold", size=15),
        axis.text.x = element_text(colour = "black", size=12),
        axis.text.y = element_text(colour = "black", size=12), 
        legend.title = element_text(size=11.5, face = "bold"),
        legend.text = element_text(size=11.5),
        legend.position="none")#http://127.0.0.1:24155/graphics/plot_zoom_png?width=1762&height=934

ev1 <- expression(paste(R^2, "0.3831195 Robust Clade"))
ev2 <- expression(paste(m, "0.004940128 Robust Clade"))
p_neuM_seawater1 = p_neuM_seawater + annotate("text", x = -3.5, y=0.98, size=6.5, label=ev1, 
                                              fontface="bold", color="black",
                                              parse=TRUE) +
annotate("text", x = -3.5, y=0.9, size=6.5, label=ev2, 
           fontface="bold", color="black",
           parse=TRUE)

getwd()

mp_ncm = ggdraw() +
  draw_plot(p_neuM_coral1, x = 0.01, y = 0.1, width = 0.49, height = 0.8) + 
  draw_plot(p_neuM_seawater1, x = 0.51, y = 0.1, width = 0.49, height = 0.8) 
mp_ncm
ggsave("neuModel.jpg", mp_ncm, width = 20, height = 10, dpi= 600)


```

## Neutral Model Aus
```{r}
####NCM Neutral model############################

#rm(list=ls())#clear previous data
#Adam Burns - 2/10/2015
#aburns2@uoregon.edu

#From Burns et al. Contribution of neutral processes to the assembly of the gut microbial communities changes over host development

#Fits the neutral model from Sloan et al. 2006 to an OTU table and returns several fitting statistics. Alternatively, will return predicted occurrence frequencies for each OTU based on their abundance in the metacommunity when stats=FALSE. For use in R.

#spp: A community table for communities of interest with local communities/samples as rows and taxa as columns. All samples must be rarefied to the same depth
#coral <- subset_samples(gcmp.rarefied, Biological_group == "yes" & coral_group =="yes") # All with outgroups millepora and stylerasteridae 


##Sub_sample Corals Acropora
acro <-subset_samples(ch_phylo, country == "Hawaii" & family == "Acroporidae")
acro
set.seed(111) # keep result reproductive
acro.rarefied = rarefy_even_depth(acro, rngseed=1, sample.size=3000, replace=T)
acro.rarefied
acro

## Subsample Meru
meru <-subset_samples(ch_phylo, country == "Hawaii" & family == "Pocilloporidae")
set.seed(111) # keep result reproductive
meru.rarefied = rarefy_even_depth(meru, rngseed=1, sample.size=3000, replace=T)
meru.rarefied
meru
## Environmental Samples 
env_sample <- subset_samples(ch_phylo, Biological_group == "no" & country == "Hawaii")
env_sample

set.seed(111) # keep result reproductive
env.rarefied = rarefy_even_depth(env_sample, rngseed=1, sample.size=3000, replace=T)
env.rarefied
env_sample

##### Subsample complex clade
complex <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "complex")

set.seed(111)
complex.rarefied = rarefy_even_depth(complex, rngseed=1, sample.size=3000, replace=T)
complex.rarefied
complex


## Subsample complex mucus
complex.m <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "complex" & BiologicalMatter == "Coral Mucus")

set.seed(111)
complex.m.rarefied = rarefy_even_depth(complex.m, rngseed=1, sample.size=3000, replace=T)
complex.m.rarefied
complex.m

## Subsample complex tissue
complex.s <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "complex"& BiologicalMatter == "Coral Skeleton")

set.seed(111)
complex.s.rarefied = rarefy_even_depth(complex.s, rngseed=1, sample.size=3000, replace=T)
complex.s.rarefied
complex.s

##### Subsample Robust clade
robust <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "robust")

set.seed(111)
robust.rarefied = rarefy_even_depth(robust, rngseed=1, sample.size=3000, replace=T)
robust.rarefied
robust


## Subsample robust
robust.m <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "robust" & BiologicalMatter == "Coral Mucus")

set.seed(111)
robust.m.rarefied = rarefy_even_depth(robust.m, rngseed=1, sample.size=3000, replace=T)
robust.m.rarefied
robust

## Subsample robust
robust.s <-subset_samples(ch_phylo, country == "Hawaii" & complex_robust == "robust"& BiologicalMatter == "Coral Skeleton")

set.seed(111)
robust.s.rarefied = rarefy_even_depth(robust.s, rngseed=1, sample.size=3000, replace=T)
robust.s.rarefied
robust.s
#pool: A community table for defining source community (optional; Default=NULL).

#taxon: A table listing the taxonomic calls for each otu, with OTU ids as row names and taxonomic classifications as columns.

#If stats=TRUE the function will return fitting statistics.
#If stats=FALSE the function will return a table of observed and predicted values for each otu.

## Plot distribution
glom_meru <- tax_glom(meru.rarefied, taxrank = 'Family', NArm = TRUE)
glom_acro <- tax_glom(acro.rarefied, taxrank = 'Family', NArm = TRUE)
glom_env <- tax_glom(env.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp <- tax_glom(complex.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob <- tax_glom(robust.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp_m <- tax_glom(complex.m.rarefied, taxrank = 'Family', NArm = TRUE)
glom_comp_s <- tax_glom(complex.s.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob_m <- tax_glom(robust.m.rarefied, taxrank = 'Family', NArm = TRUE)
glom_rob_s <- tax_glom(robust.s.rarefied, taxrank = 'Family', NArm = TRUE)



########## define sncm.fit ###########################
sncm.fit <- function(spp, pool=NULL, stats=TRUE, taxon=NULL){
	require(minpack.lm)
	require(Hmisc)
	require(stats4)
	
	options(warn=-1)

	#Calculate the number of individuals per community
	N <- mean(apply(spp, 1, sum))
	
	#Calculate the average relative abundance of each taxa across communities
	if(is.null(pool)){
		p.m <- apply(spp, 2, mean)
		p.m <- p.m[p.m != 0]
		p <- p.m/N
	} else {
		p.m <- apply(pool, 2, mean)
		p.m <- p.m[p.m != 0]
		p <- p.m/N
	}

#Calculate the occurrence frequency of each taxa across communities
spp.bi <- 1*(spp>0)
freq <- apply(spp.bi, 2, mean)
freq <- freq[freq != 0]
#Combine
C <- merge(p, freq, by=0)
C <- C[order(C[,2]),]
C <- as.data.frame(C)
C.0 <- C[!(apply(C, 1, function(y) any(y == 0))),] #Removes rows with any zero (absent in either source pool or local communities)
p <- C.0[,2]
freq <- C.0[,3]
names(p) <- C.0[,1]
names(freq) <- C.0[,1]

#Calculate the limit of detection
d = 1/N

##Fit model parameter m (or Nm) using Non-linear least squares (NLS)
m.fit <- nlsLM(freq ~ pbeta(d, N*m*p, N*m*(1-p), lower.tail=FALSE), start=list(m=0.1))
m.ci <- confint(m.fit, 'm', level=0.95)
	
##Fit neutral model parameter m (or Nm) using Maximum likelihood estimation (MLE)
sncm.LL <- function(m, sigma){
	R = freq - pbeta(d, N*m*p, N*m*(1-p), lower.tail=FALSE)
	R = dnorm(R, 0, sigma)
	-sum(log(R))
}
	
m.mle <- mle(sncm.LL, start=list(m=0.1, sigma=0.1), nobs=length(p))

##Calculate Akaike's Information Criterion (AIC)
aic.fit <- AIC(m.mle, k=2)
bic.fit <- BIC(m.mle)

##Calculate goodness-of-fit (R-squared and Root Mean Squared Error)
freq.pred <- pbeta(d, N*coef(m.fit)*p, N*coef(m.fit)*(1-p), lower.tail=FALSE)
Rsqr <- 1 - (sum((freq - freq.pred)^2))/(sum((freq - mean(freq))^2))
RMSE <- sqrt(sum((freq-freq.pred)^2)/(length(freq)-1))
	
pred.ci <- binconf(freq.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)
	
##Calculate AIC for binomial model
bino.LL <- function(mu, sigma){
R = freq - pbinom(d, N, p, lower.tail=FALSE)
R = dnorm(R, mu, sigma)
-sum(log(R))
	}
bino.mle <- mle(bino.LL, start=list(mu=0, sigma=0.1), nobs=length(p))
	
aic.bino <- AIC(bino.mle, k=2)
bic.bino <- BIC(bino.mle)
	
##Goodness of fit for binomial model
bino.pred <- pbinom(d, N, p, lower.tail=FALSE)
Rsqr.bino <- 1 - (sum((freq - bino.pred)^2))/(sum((freq - mean(freq))^2))
RMSE.bino <- sqrt(sum((freq - bino.pred)^2)/(length(freq) - 1))

bino.pred.ci <- binconf(bino.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)
	
##Calculate AIC for Poisson model
pois.LL <- function(mu, sigma){
	R = freq - ppois(d, N*p, lower.tail=FALSE)
	R = dnorm(R, mu, sigma)
	-sum(log(R))
}
pois.mle <- mle(pois.LL, start=list(mu=0, sigma=0.1), nobs=length(p))
	
aic.pois <- AIC(pois.mle, k=2)
bic.pois <- BIC(pois.mle)
	
##Goodness of fit for Poisson model
pois.pred <- ppois(d, N*p, lower.tail=FALSE)
Rsqr.pois <- 1 - (sum((freq - pois.pred)^2))/(sum((freq - mean(freq))^2))
RMSE.pois <- sqrt(sum((freq - pois.pred)^2)/(length(freq) - 1))

pois.pred.ci <- binconf(pois.pred*nrow(spp), nrow(spp), alpha=0.05, method="wilson", return.df=TRUE)

##Results
	if(stats==TRUE){
		fitstats <- data.frame(m=numeric(), m.ci=numeric(), m.mle=numeric(), maxLL=numeric(), binoLL=numeric(), poisLL=numeric(), Rsqr=numeric(), Rsqr.bino=numeric(), Rsqr.pois=numeric(), RMSE=numeric(), RMSE.bino=numeric(), RMSE.pois=numeric(), AIC=numeric(), BIC=numeric(), AIC.bino=numeric(), BIC.bino=numeric(), AIC.pois=numeric(), BIC.pois=numeric(), N=numeric(), Samples=numeric(), Richness=numeric(), Detect=numeric())
		fitstats[1,] <- c(coef(m.fit), coef(m.fit)-m.ci[1], m.mle@coef['m'], m.mle@details$value, bino.mle@details$value, pois.mle@details$value, Rsqr, Rsqr.bino, Rsqr.pois, RMSE, RMSE.bino, RMSE.pois, aic.fit, bic.fit, aic.bino, bic.bino, aic.pois, bic.pois, N, nrow(spp), length(p), d)
		return(fitstats)
	} else {
	A <- cbind(p, freq, freq.pred, pred.ci[,2:3], bino.pred, bino.pred.ci[,2:3])
	A <- as.data.frame(A)
		colnames(A) <- c('p', 'freq', 'freq.pred', 'pred.lwr', 'pred.upr', 'bino.pred', 'bino.lwr', 'bino.upr')
		if(is.null(taxon)){
			B <- A[order(A[,1]),]
		} else {
			B <- merge(A, taxon, by=0, all=TRUE)
			row.names(B) <- B[,1]
			B <- B[,-1]
			B <- B[order(B[,1]),]
		}
		return(B)
	}
}
##############################
#library("phyloseq")
#otu <- read.delim('otu_whole.txt', head = T, row.names = 1, sep = '\t', stringsAsFactors = FALSE, check.names = FALSE)
#sampledata <- read.delim('sampledata.txt', head = T, row.names = 1, sep = '\t', stringsAsFactors = FALSE, check.names = FALSE)

#otu <- t(otu)
#OTU = otu_table(spp, taxa_are_rows = FALSE)
#sampledata =sample_data(sampledata)
#tree <- read.tree("rep_set.tre",tree.names = NULL)

#neutral= phyloseq(OTU, tree, sampledata)
#neutral
#### 01/16/2024: At this stage, I have already completed steps above and I used the phyloseq object used above to start removing singletons


##Sub-sampling to equal sequening efforts
#physeq2 = rarefy_even_depth(neutral.nonsing, rngseed = 12345,
 #                           sample.size = 42110,
 #                           replace = F)
set.seed(111)
###coral communities
coral <- prune_taxa(taxa_sums(glom_acro) > 0, glom_acro)
coral
otu_coral = as.data.frame(t(otu_table(coral)))

mod_w_coral = sncm.fit(otu_coral)
mod1_w_coral = sncm.fit(otu_coral,stats = F)

##Identify above, below and neutral OTUs 
addf = function(freq, pred.upr, pred.lwr) {
  if(freq > pred.upr){
    model = "above"
  }
  else if(freq < pred.lwr){
    model = "below"
  }
  else if(pred.upr >= freq & freq >= pred.lwr){
    model = "neutral"
  }
  return(model)
}

mod1_w_coral$model = mapply(addf, mod1_w_coral$freq, mod1_w_coral$pred.upr, mod1_w_coral$pred.lwr)  
length(which(mod1_w_coral$model == "above"))
length(which(mod1_w_coral$model == "below"))
length(which(mod1_w_coral$model == "neutral"))

mod1_w_coral$model <- factor(mod1_w_coral$model, 
                             levels = c("above", "neutral", "below"))
#coral_model_adj_taxa = data.frame(adjustp = mod1_w_coral$p,
#                    adjustp = sum(mod(phylo_coral),
#                    tax_table(phylo_coral))

## Assign Taxonomy to table of significant microbes
coral_taxa = cbind(as(mod1_w_coral, "data.frame"), as(tax_table(glom_acro)[rownames(mod1_w_coral), ], "matrix"))                    

#plot neuModel
p_neuM_coral=ggplot(mod1_w_coral, aes(x=log10(p), y=freq, color = model))+
  geom_point(shape=19, alpha=0.7, size=2.5)+
  geom_line(aes(x=log10(p), y=freq.pred),
            color="blue",linetype="solid",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.lwr),
            color="blue",linetype="dashed",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.upr),
            color="blue",linetype="dashed",linewidth=1)+
  scale_colour_manual(name="OTUs",
                      values=c("#E41A1C", "#999999", "#FF7F00"),
                      breaks=c("above", "neutral", "below"),
                      labels=c("Over-represented", "Neutrally-distributed", "Under-represented"))+
  xlab("Log10 (Mean Relative abundance)") + ylab("Occurance frequency")+
  theme_bw() +
  theme(axis.title.x = element_text(face="bold", size=15),
        axis.title.y = element_text(face="bold", size=15),
        axis.text.x = element_text(colour = "black", size=12),
        axis.text.y = element_text(colour = "black", size=12), 
        legend.title = element_text(size=11.5, face = "bold"),
        legend.text = element_text(size=11.5),
        legend.position="none")


lb1 <- mod_w_coral$Rsqr
lb2 <- mod_w_coral$m
lb3 <- expression(paste("Hawaii Acroporidae\n Coral Samples"))
p_neuM_coral1 = p_neuM_coral + annotate("text", x = -4.5, y=0.98, size=3.5, label=lb1, 
                                        fontface="bold", color="black",
                                        parse=TRUE) +
  annotate("text", x = -4.5, y=0.9, size=3.5, label=lb2, 
           fontface="bold", color="black",
           parse=TRUE) + annotate("text", x=-3.0, y=0.9, size=6.5, label=lb3, fontface="bold", color="black", parse=TRUE)

set.seed(111)
###seawater communities
#seawater <- subset_samples(physeq2, SampleType == "seawater")
seawater <- prune_taxa(taxa_sums(glom_env) > 0, glom_env)
#seawater

otu_seawater = as.data.frame(t(otu_table(seawater)))

mod_w_seawater = sncm.fit(otu_seawater)
mod1_w_seawater = sncm.fit(otu_seawater,stats = F)

##Identify above, below and netural OTUs 
addf = function(freq, pred.upr, pred.lwr) {
  if(freq > pred.upr){
    model = "above"
  }
  else if(freq < pred.lwr){
    model = "below"
  }
  else if(pred.upr >= freq & freq >= pred.lwr){
    model = "neutral"
  }
  return(model)
}

mod1_w_seawater$model = mapply(addf, mod1_w_seawater$freq, mod1_w_seawater$pred.upr, mod1_w_seawater$pred.lwr)  
length(which(mod1_w_seawater$model == "above"))
length(which(mod1_w_seawater$model == "below"))
length(which(mod1_w_seawater$model == "neutral"))

mod1_w_seawater$model <- factor(mod1_w_seawater$model, 
                                levels = c("above", "neutral", "below"))


## Assign Taxonomy to table of significant microbes
env_taxa = cbind(as(mod1_w_seawater, "data.frame"), as(tax_table(glom_env)[rownames(mod1_w_seawater), ], "matrix"))



#plot neuModel
p_neuM_seawater=ggplot(mod1_w_seawater, aes(x=log10(p), y=freq, color = model))+
  geom_point(shape=19, alpha=0.7, size=2.5)+
  geom_line(aes(x=log10(p), y=freq.pred),
            color="blue",linetype="solid",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.lwr),
            color="blue",linetype="dashed",linewidth=1)+
  geom_line(aes(x=log10(p), y=pred.upr),
            color="blue",linetype="dashed",linewidth=1)+
  scale_colour_manual(name="OTUs",
                      values=c("#E41A1C", "#999999", "#FF7F00"),
                      breaks=c("above", "neutral", "below"),
                      labels=c("Over-represented", "Neutrally-distributed", "Under-represented"))+
  xlab("Log10 (Mean Relative abundance)") + ylab("Occurance frequency")+
  theme_bw() +
  theme(axis.title.x = element_text(face="bold", size=15),
        axis.title.y = element_text(face="bold", size=15),
        axis.text.x = element_text(colour = "black", size=12),
        axis.text.y = element_text(colour = "black", size=12), 
        legend.title = element_text(size=11.5, face = "bold"),
        legend.text = element_text(size=11.5),
        legend.position="none")#http://127.0.0.1:24155/graphics/plot_zoom_png?width=1762&height=934

ev1 <- mod_w_seawater$Rsqr
ev2 <- mod_w_seawater$m
ev3 <- expression(paste("Hawaii \n Environmental Samples"))
p_neuM_seawater1 = p_neuM_seawater + annotate("text", x = -4.5, y=0.98, size=3.5, label=ev1, 
                                        fontface="bold", color="black",
                                        parse=TRUE) +
  annotate("text", x = -4.5, y=0.9, size=3.5, label=ev2, 
           fontface="bold", color="black",
           parse=TRUE) + annotate("text", x=-3.0, y=0.9, size=6.5, label=ev3, fontface="bold", color="black", parse=TRUE)

getwd()

mp_ncm = ggdraw() +
  draw_plot(p_neuM_coral1, x = 0.01, y = 0.1, width = 0.49, height = 0.8) + 
  draw_plot(p_neuM_seawater1, x = 0.51, y = 0.1, width = 0.49, height = 0.8) 
mp_ncm
ggsave("neuModel.jpg", mp_ncm, width = 20, height = 10, dpi= 600)


```


#Null Models
```{r}
##################null model##################

#######bNTI
library(picante)
library(ape)
library(parallel)

otu <- read.delim(file="otu_seawater.txt", sep = "\t", header = TRUE, row.names=1)
tree <- read.tree("rep_set.tre",tree.names = NULL)
rand <- 1000

####match data
matched <- match.phylo.comm(tree,t(otu))
comm<- matched$comm
phylo <- matched$phy

####computes the cophenetic distances 
# The cophenetic distance between two objects is the height of the dendrogram where the two branches that include the two objects merge into a single branch

phydist <- cophenetic(phylo)
weighted = TRUE

####calculate the observed βMNTD
bmntd.obs <- as.matrix(comdistnt(comm,phydist,abundance.weighted = weighted,exclude.conspecifics = FALSE))

####define βMNTD.randomization function
bMNTD.randomization <- function(i, dis, com, weighted, exclude.consp) {
  library(picante)
  dist.rand = dis
  rand.name = sample(colnames(dis))
  colnames(dist.rand) = rand.name
  rownames(dist.rand) = rand.name
  bMNTD.rand <- as.matrix(comdistnt(com,dist.rand, abundance.weighted = weighted,exclude.conspecifics = exclude.consp))
  bMNTD.rand
}

####recalculate the values of βMNTD by randomization 1000 times
c1 <- parallel::makeCluster(2, type = "PSOCK")
bmntd.random <- parallel::parLapply(c1, 1:rand, bMNTD.randomization,
                                    dis = phydist, com = comm, weighted = weighted, exclude.consp = FALSE)
parallel::stopCluster(c1)

####calculate the βNTI
bmntd.random.array <- array(unlist(bmntd.random),dim = c(nrow(bmntd.random[[1]]),ncol(bmntd.random[[1]]),length(bmntd.random)))
bNTI = (bmntd.obs - apply(bmntd.random.array, c(1, 2), mean))/(apply(bmntd.random.array,c(1, 2), sd))
diag(bNTI) <- 0
bNTI[is.na(bNTI)] = 0
bNTI[is.na(bmntd.obs)] = NA

write.csv(bNTI,paste("bNTI",weighted,"weighted1.csv",sep = "."))

###################RCbray
comm.orig <- read.delim(file="otu_coral.txt", sep = "\t", header = TRUE, row.names=1)
comm.trans <- t(comm.orig)

####calculate the number of shared species (SSobs)
method="bray"
if(method=="jaccard"){
  binary=TRUE
}else{
  binary=FALSE
}
comm <- comm.trans[,colSums(comm.trans)>0]
dt.obs <- as.matrix(vegdist(comm,method = method,binary = binary))

####calculate the distribution of the expected shared species from null model (SSexp)
dt.null.list <- list()
rand = 1000

for (i in 1:rand) {
  comm.rand <- comm
  comm.rand[]=0
  prob.species <- colSums(comm>0)
  prob.abundance <- colSums(comm)
  species <- rowSums(comm>0)
  samps <- rowSums(comm)
  for (j in 1:nrow(comm)) {
    id.species <- sample(1:ncol(comm), species[j], replace = FALSE, prob = prob.species)
    count <- sample(id.species, (samps[j] - species[j]), replace = TRUE,prob = prob.abundance[id.species])
    table <- table(count)
    comm.rand[j, as.numeric(names(table))] = as.vector(table)
    comm.rand[j, id.species] = comm.rand[j, id.species] + 1
  }
  
  dt.null.list[[i]] = as.matrix(vegdist(comm.rand, method = method,binary = binary))
}

dt.null.array <- array(unlist(dt.null.list),dim=c(nrow(dt.null.list[[1]]),ncol(dt.null.list[[1]]),length(dt.null.list)))

####compare SSobs with the distribution of SSexp and calculate the RCbray
comp <- function(x, c) {
  (x < c) + 0.5 * (x == c)
}

alpha = matrix(rowSums(apply(dt.null.array,3,comp,c=dt.obs)),nrow = nrow(dt.obs))/rand
rc = (alpha - 0.5) * 2
rownames(rc) = rownames(dt.obs)
colnames(rc) = colnames(dt.obs)
write.csv(rc,paste("coral_RaupCrick.",method,".",rand,".csv",sep = ""))

####summary_NTI_RC

####determine the different ecological progresses according the values of RCbray and βNTI
bNTI <- read.csv("bNTI.TRUE.weighted.csv",row.names = 1)
RC <- read.csv("coral_RaupCrick.bray.csv",row.names = 1)

samp.nti <- names(bNTI)
samp.rc <- names(RC)
matched <- match(samp.nti,samp.rc)
RC <- RC[matched,matched]

bNTI.matrix <- as.matrix(bNTI)
rowname <- rownames(bNTI.matrix)
colname <- colnames(bNTI.matrix)
rown <- row(bNTI.matrix)
coln <- col(bNTI.matrix)
bNTI.v <- as.vector(as.dist(bNTI.matrix))
RC.v <- as.vector(as.dist(as.matrix(RC)))
rown.v <- as.vector(as.dist(rown))
coln.v <- as.vector(as.dist(coln))
result <- data.frame(name1 <- rowname[rown.v],name2 <- colname[coln.v],
                     bNTI <- bNTI.v,RC <- RC.v)
colnames(result) <- c("name1","name2","bNTI","RC")

eco.pro <- function(res, sig.nti, sig.rc){
  nti <- res[1]
  rc <- res[2]
  if(nti > sig.nti){
    process <- "Heterogeneous selection"
  }else if(nti < (-sig.nti)){
    process <- "Homogeneous selection"
  }else if(abs(nti) < sig.nti && rc < (-sig.rc) ){
    process <- "Homogenizing dispersal"
  }else if(abs(nti) < sig.nti && rc > sig.rc ){
    process <- "Dispersal limitation"
  }else{
    process <- "Undominated"
  }
  process
}

result[,5]<-sapply(as.list(as.data.frame(t(result[,3:4]))),eco.pro,sig.nti=2,sig.rc=0.95)
colnames(result)[5] <- "EcologicalProcess"
write.csv(result,"Ecological process bNTI RC.csv")

```

# DISTANCE MATRICES
```{r}
## Main Coral 
bray_coral = phyloseq::distance(main_coral, method="bray") ## divergent age pattern
uni_coral = phyloseq::distance(main_coral, method="unifrac") ## divergent age pattern
wuni_coral = phyloseq::distance(main_coral, method="wUnifrac") ## divergent age pattern

coral_df <- data.frame(sample_data(main_coral))

####### Australia ######
##Distance matrix for complex Australia complex
bray_complex_aus_tis = phyloseq::distance(complex_aus_tis, method="bray") ## divergent age pattern
uni_complex_aus_tis = phyloseq::distance(complex_aus_tis, method="unifrac")  ## divergent age pattern
wuni_complex_aus_tis = phyloseq::distance(complex_aus_tis, method="wUnifrac") ## no pattern

##Distance matrix for complex Australia acropora genus ## there was only one species in tissue so I switched to skeleton
bray_acropora_aus_ske = phyloseq::distance(acropora_aus_ske, method="bray") ## 
uni_acropora_aus_ske = phyloseq::distance(acropora_aus_ske, method="unifrac")  ## 
wuni_acropora_aus_ske = phyloseq::distance(acropora_aus_ske, method="wUnifrac") ## 

##Distance matrix for complex Australia Pachyseris
bray_pachyseris_aus_tis = phyloseq::distance(Pachyseris_aus_tis, method="bray") ## 
uni_pachyseris_aus_tis = phyloseq::distance(Pachyseris_aus_tis, method="unifrac")  ## 
wuni_pachyseris_aus_tis = phyloseq::distance(Pachyseris_aus_tis, method="wUnifrac") ## 

##Distance matrix for complex Australia robust
bray_robust_aus_tis = phyloseq::distance(robust_aus_tis, method="bray") ## uneven sampling makes it hard to observe trends with Diploastrea
uni_robust_aus_tis = phyloseq::distance(robust_aus_tis, method="unifrac")
wuni_robust_aus_tis = phyloseq::distance(robust_aus_tis, method="wUnifrac")

# so instead I focused on merulinidae Unfortunately, hard to identify trends with low sample size.
bray_merulinidae_aus_tis = phyloseq::distance(merulinidae_aus_tis, method="bray")
uni_merulinidae_aus_tis = phyloseq::distance(merulinidae_aus_tis, method="unifrac")
wuni_merulinidae_aus_tis = phyloseq::distance(merulinidae_aus_tis, method="wUnifrac")

####### Caribbean ######
##Distance matrix for Caribbean complex ## again uneven sampling makes it hard to compare wtihin complex clade acropora dominates 
bray_complex_carib_tis = phyloseq::distance(complex_carib_tis, method="bray") 
bray_complex_carib_tis = phyloseq::distance(complex_carib_tis, method="unifrac") 
bray_complex_carib_tis = phyloseq::distance(complex_carib_tis, method="wUnifrac") 
## acropora 
bray_acroporidae_cairb_tis = phyloseq::distance(acroporidae_cairb_tis, method="bray") 
uni_acroporidae_cairb_tis = phyloseq::distance(acroporidae_cairb_tis, method="unifrac")
wuni_acroporidae_cairb_tis = phyloseq::distance(acroporidae_cairb_tis, method="wUnifrac")


##Distance matrix for Caribbean robust 
bray_robust_carib_tis = phyloseq::distance(robust_carib_tis, method="bray")
uni_robust_carib_tis = phyloseq::distance(robust_carib_tis, method="unifrac")
wuni_robust_carib_tis = phyloseq::distance(robust_carib_tis, method="wUnifrac")

##Distance matrix for Caribbean robust skeleton 
bray_robust_carib_ske = phyloseq::distance(robust_carib_ske, method="bray")
uni_robust_carib_ske = phyloseq::distance(robust_carib_ske, method="unifrac")
wuni_robust_carib_ske = phyloseq::distance(robust_carib_ske, method="wUnifrac")

##Distance matrix for Caribbean Orbicella Tissue 
bray_orbicella_tis = phyloseq::distance(orbicella_tis, method="bray")
uni_orbicella_tis = phyloseq::distance(orbicella_tis, method="unifrac")
wuni_orbicella_tis = phyloseq::distance(orbicella_tis, method="wUnifrac")

##Distance matrix for Caribbean Orbicella Skeleton 
bray_orbicella_ske = phyloseq::distance(orbicella_ske, method="bray")
uni_orbicella_ske = phyloseq::distance(orbicella_ske, method="unifrac")
wuni_orbicella_ske = phyloseq::distance(orbicella_ske, method="wUnifrac")

####### Hawaii ######
##Distance matrix for Hawaii complex tissue analysis
bray_complex_hawaii_tis = phyloseq::distance(complex_hawaii_tis, method="bray")
uni_complex_hawaii_tis = phyloseq::distance(complex_hawaii_tis, method="unifrac")
wuni_complex_hawaii_tis = phyloseq::distance(complex_hawaii_tis, method="wUnifrac")

##Distance matrix for Hawaii Robust  tissue analysis
bray_robust_hawaii_tis = phyloseq::distance(robust_hawaii_tis, method="bray")
uni_robust_hawaii_tis = phyloseq::distance(robust_hawaii_tis, method="unifrac")
wuni_robust_hawaii_tis = phyloseq::distance(robust_hawaii_tis, method="wUnifrac")

##Distance matrix for complex phylosymbiosis analysis
bray_pavona_tis = phyloseq::distance(pavona_tis, method="bray")
uni_pavona_tis = phyloseq::distance(pavona_tis, method="unifrac")
wuni_pavona_tis = phyloseq::distance(pavona_tis, method="wUnifrac")

##Distance matrix for complex phylosymbiosis analysis
bray_pavona_ske = phyloseq::distance(pavona_ske, method="bray")
uni_pavona_ske = phyloseq::distance(pavona_ske, method="unifrac")
wuni_pavona_ske = phyloseq::distance(pavona_ske, method="wUnifrac")

```

# Ordination from Distance Matrix and plot PCoA 
```{r}
##Calculate the PCoA on Complex Corals
rt.pcoa = ordinate(coral, method="PCoA", distance=wuni_coral)
# Set variables to zero for subsequent changes
#Ofra
pcoa<-0
pcoa <- plot_ordination(coral, rt.pcoa,  shape = "country_region",color="family")+  geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=country_region))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+
                        

##Calculate the PCoA on Robust Corals
rt.pcoa1 = ordinate(robust_ske, method="PCoA", distance=bray_rob_ske)
pcoa1<-0
pcoa1 <- plot_ordination(robust_ske, rt.pcoa1, shape = "country_region", color="family")+  geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=family))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+
                        
##Calculate the PCoA on meru Corals
rt.pcoa2 = ordinate(meru_tis, method="PCoA", distance=uni_meru_tis)
pcoa2<-0
pcoa2 <- plot_ordination(meru_tis, rt.pcoa2, shape = "country_region", color="host_genus") + scale_color_brewer(palette="Paired") + geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=host_genus))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+

##Calculate the PCoA on Pocilloporidae Corals
rt.pcoa3 = ordinate(poci_ske, method="PCoA", distance=bray_poci_ske)
pcoa3<-0
pcoa3 <- plot_ordination(poci_ske, rt.pcoa3, shape = "host_genus", color="country") + geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=country))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+

##Calculate the PCoA on Acroporidae Corals
rt.pcoa4 = ordinate(acro_tis, method="PCoA", distance=bray_acro_tis)
pcoa4<-0
pcoa4 <- plot_ordination(acro_tis, rt.pcoa4, shape = "country_region", color="host_genus")+  geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=host_genus))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+


##Calculate the PCoA on Agariciidae Corals
rt.pcoa5 = ordinate(agra_tis, method="PCoA", distance=uni_agra_tis)
pcoa5<-0
pcoa5 <- plot_ordination(agra_tis, rt.pcoa5, shape = "country_region", color="host_genus")+ scale_fill_manual(name=NULL,
                    values =c("#B2DF8A","#E31A1C","#FB9A99","#6A3D9A")) +  geom_point(size=3) + theme_bw() + theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=host_genus))+ theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) # title =expression(paste(italic(" Orbicella spp")," Bray-Curtis Distance PCoA across Depth")) scale_color_manual(values=c("#E31A1C","#1F78B4"))+

```

# Alpha Diversity BoxPlots Compartments & Family
```{r, boxplots}
###Once we have our sub_samples rarefied we can start with some diversity analysis  and visualization
### Alpha Diversity ** you can make multiple diversity box plots by switch x for sample variables
#install.packages("adespatial")
#install_github("umerijaz/microbiomeSeq") # Install the package

#library(adespatial)
#library(microbiomeSeq)  #load the package

alpha_family_mucus <- plot_richness(mucus, x="family", color = "complex_robust", title = "Mucus Diversity Across family", measures = ("Shannon"),)+
  geom_boxplot(width=0.8, fill="white")  + 
  scale_fill_manual(name=NULL,
                    values = c(brewer.pal(12,"Paired"))) +
    theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

alpha_family_tissue <- plot_richness(tissue, x="family",color = "complex_robust", title = "Tissue Diversity Across family", color = "family", measures = ("Shannon"),)+
  geom_boxplot(width=0.8, fill="white")+
    theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

alpha_family_skeleton <- plot_richness(skeleton, x="family",color = "complex_robust", title = "Skeleton Diversity Across family", color = "family", measures = ("Shannon"),)+
  geom_boxplot(width=0.8, fill="white")+
    theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

#all_df$tissue_compartment

complex_diversity_compartment <- plot_richness(complex, x="family", color = "Genus", title = "Merulinidae Diversity Across compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))
 
robust_diversity_compartment <- plot_richness(robust, x="family",color = "Genus", title = "Alpha Diversity Across Compartment in Panama", measures = ("Shannon"),)+
  geom_boxplot(width=0.8, fill="white")+
    theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

acro_diversity_compartment <- plot_richness(acro, x="country", title = "Acroporidae Diversity Across Species Tissue", color = "country",measures = ("Shannon"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

meru_compartment <- plot_richness(meru, x="country", color = "Genus",title = "Merulinidae Diversity Across Species Tissue", color = "country", measures = ("Shannon"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

agra_compartment <- plot_richness(agra, x="country",color = "Genus", title = "Merulinidae Diversity Across Species Tissue", color = "country",measures = ("Shannon"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

poli_compartment <- plot_richness(poci, x="country", color = "Genus",title = "Merulinidae Diversity Across Species Tissue", color = "country",measures = ("Shannon"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

# 
```


# BAR PLOTS: Coral Compartments
```{r, message=FALSE, warning=FALSE}
## Bar Plots Across Family and biological compartment
phylo_comp <-subset_samples(phylo_coral, coral_group == "yes" & family_group == "yes" & complex_robust == "robust")# & BiologicalMatter == "Coral Skeleton")

phylo_comp <- subset_taxa(phylo_comp, Order !="Unclassified Bacteroidia")
# agglomerate taxa
glom <- tax_glom(phylo_comp, taxrank = 'Order', NArm = TRUE)

# Transform sample counts to proportion out of 100
ps.all = transform_sample_counts(glom, function(x) x/sum(x)*100)
ps.melt <- psmelt(ps.all)

# change to character for easy-adjusted level
ps.melt <- ps.melt %>%
  group_by(family,BiologicalMatter,host_genus,country_region,Order) %>%
  mutate(median=median(Abundance)) %>%
  arrange(desc(median))

# select group median > 1
keep <- unique(ps.melt$Order[ps.melt$median > 20])
ps.melt$Order[!(ps.melt$Order %in% keep)] <- "< %2 (Other)"

#to get the same rows together
ps.melt_sum <- ps.melt %>%
  group_by(family,BiologicalMatter,host_genus,country_region,Order) %>%
  summarize(Abundance = sum(Abundance), 
            mean = max(Abundance), .groups = "drop") %>%
            mutate(Order= factor(Order),
                   Order= fct_reorder(Order,mean, .desc = TRUE),
                   Order= fct_shift(Order, n=1))

   
#, .desc = TRUE 
# Then Plot Results
many_col <- c("#A6CEE3", "#33A02C", "#B2DF8A",  "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00","#FDBF6F", "#FB9A99", "#CAB2D6",  "#B15928","#1B9E77", "#D95F02", "#FFFF99","#E7298A", "#7570B3","#666666" ,"#66A61E", "#E6AB02","#A6761D", "#E41A1C", "#377EB8", "#4DAF4A" ,"#984EA3", "#FF7F00", "#FFFF33", "#A65628", "#F781BF" ,"#D53E4F","#F46D43", "#FDAE61", "#FEE08B","#999999", "#E6F598","#ABDDA4","#3288BD","cyan1","#66C2A5")

testp <-0
testp <-ggplot(ps.melt_sum, aes(x = family, y = Abundance, fill = Order)) + 
  geom_bar(stat = "identity", position="fill")  +
  labs( x=" family", y="Percent Relative Abundance (%)") +  scale_fill_manual(name=NULL, values = c(many_col))  +  facet_wrap(~BiologicalMatter, scales= "free_x", nrow=1) +  
  theme_classic() + theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = 180) +
  theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold"), plot.title = element_text(size = 14, face = "bold",hjust = 0.5,),legend.text = element_text(size = 8), legend.box = "vertical") + theme(legend.text = element_text(size=12))+
                  theme(legend.key.size = unit(.9, 'cm')))  #title =expression(paste(italic(" Oribcella Spp."),"Microbial Community Composition Across Depth ")), + #
plot(testp)

##################################################################
## Coral Mucus and Environmental Samples

## Pcoa analysis showed robust tissue structured by site
## Bar Plots Across Family and biological compartment
mucus_comp <-subset_samples(coral, complex_robust == "robust" & BiologicalMatter == "Coral Tissue")

ps.all2 = transform_sample_counts(mucus_comp, function(x) x/sum(x)*100)

# agglomerate taxa
glom2 <- tax_glom(ps.all2, taxrank = 'Order', NArm = TRUE)
ps.melt2 <- psmelt(glom2)

# change to character for easy-adjusted level
ps.melt2 <- ps.melt2 %>%
  group_by(family,BiologicalMatter,host_genus,country_region,Order) %>%
  mutate(median=median(Abundance)) %>%
  arrange(desc(median))

# select group median > 1
keep2 <- unique(ps.melt2$Order[ps.melt2$median >=5])
ps.melt2$Order[!(ps.melt2$Order %in% keep2)] <- "< %5 (Other)"

#to get the same rows together
ps.melt_sum2 <- ps.melt2 %>%
  group_by(family,BiologicalMatter,host_genus,country_region,Order) %>%
  summarize(Abundance = sum(Abundance), 
            mean = sum(Abundance), .groups = "drop") %>%
            mutate(Order= factor(Order),
                   Order= fct_reorder(Order,mean, .desc = TRUE),
                   Order= fct_shift(Order, n=-1))

   
#, .desc = TRUE 
# Then Plot Results
                  #theme(legend.key.size = unit(.9, 'cm')))  #title =expression(paste(italic(" Oribcella Spp."),"Microbial Community Composition Across Depth ")), + #
plot(testp2)

```

# BAR PLOTS: Complex Coral
```{r}
complex_coral <-subset_samples(coral, complex_robust == "robust")

## Bar Plots Across Family and biological compartment

ps.all1 = transform_sample_counts(complex_coral, function(x) x/sum(x)*100)

# agglomerate taxa
glom1 <- tax_glom(ps.all1, taxrank = 'Family', NArm = TRUE)
ps.melt_g1 <- psmelt(glom1)

# change to character for easy-adjusted level
ps.melt1 <- ps.melt_g1 %>%
  group_by(family,BiologicalMatter,country_region,Family) %>%
  mutate(median=median(Abundance)) %>%
  arrange(desc(median))

# select group median > 1
keep1 <- unique(ps.melt1$Family[ps.melt1$median >10])
ps.melt1$Family[!(ps.melt1$Family %in% keep1)] <- "< %1 (Other)"

#to get the same rows together
ps.melt_sum1 <- ps.melt1 %>%
  group_by(family,BiologicalMatter,country_region,Family) %>%
  summarize(Abundance = sum(Abundance), 
            mean = sum(Abundance), .groups = "drop") %>%
            mutate(Family= factor(Family),
                   Family= fct_reorder(Family,mean, .desc = TRUE),
                   Family= fct_shift(Family, n=1))

   
#, .desc = TRUE 
# Then Plot Results
many_col1 <- c("#A6CEE3", "#1F78B4", "#B2DF8A", "#33A02C", "#FB9A99", "#E31A1C", "#FDBF6F", "#FF7F00", "#CAB2D6", "#FFFF99", "#B15928","#1B9E77", "#D95F02", "#7570B3" ,"#666666", "#6A3D9A","#E7298A", "#66A61E", "#E6AB02", "#A6761D","#999999", "#E41A1C", "#377EB8", "#4DAF4A" ,"#984EA3", "#FF7F00", "#FFFF33", "#A65628", "#F781BF" ,"#D53E4F","#F46D43", "#FDAE61", "#FEE08B", "#E6F598","#ABDDA4","#3288BD","cyan1","#66C2A5")
testp1 <-0
testp1 <-ggplot(ps.melt_sum1, aes(x = family, y = Abundance, fill = Family)) + 
  geom_bar(stat = "identity", position="fill")  +
  labs( x=" family", y="Percent Relative Abundance (%)") +
                  scale_fill_manual(name=NULL, values = c(many_col1)) + 
                  facet_wrap(~country_region, scales= "free_x", nrow=1)+
                  theme_classic() +   theme(axis.text.x = element_text(angle = -90) +   
                  theme_classic()  +  theme(axis.text=element_text(size=12), axis.title=element_text(size=14,face="bold"), plot.title = element_text(size = 14, face = "bold",hjust = 0.5,),legend.text = element_text(size = 8), legend.box = "horizontal") + theme(legend.text = element_text(size=12))+
                  theme(legend.key.size = unit(.9, 'cm')))  #title =expression(paste(italic(" Oribcella Spp."),"Microbial Community Composition Across Depth ")), + #
plot(testp1)

```

# PERMANOVAs for Coral Compartments 
## 1. Comparison between (Mucus, Tissue, Skeleton ~ family*country_region)
```{r}
## sam_all= Mucus, Tissue and skeleton without eDNA and outgroup
# all_sam_out Tissue and skeleton with Outgroup without eDNA
##env material include mucus, tissue and skeleton, and outgroups without eDNA
distance_methods <-c("bray_coral","wuni_coral","uni_coral")
set.seed(129)
# This is what is called a for loop. I believe Kate touched on these when she gave her lecture. Try to break it down piece by piece: what is happening in the first line? The second line? The third?         
for (i in distance_methods){ 
  form <- as.formula(paste(i,"country_region+family+BiologicalMatter+sequence_run", sep="~"))
  print(form)
 adonis2(form, data=coral_df)-> result
 print(result)
}
  #capture.output(result, file = paste("C:/Users/Colin/Desktop/PSU Doctoral Program/Bio informactics - Sofia",i,"adonis_bray_all",date(),".txt"))


## Post-hoc comparison between Coral Compartments and Location
bray_coral_pairwise <-pairwise.perm.manova(bray_coral, coral_df$country_region,nperm=1000) 
uni_coral_pairwise <-pairwise.perm.manova(uni_coral,coral_df$biological_matter,nperm=1000)
wuni_coral_pairwise <-pairwise.perm.manova(wuni_coral,coral_df$biological_matter,nperm=1000)
# The above permutation shows signficant differences between biological matter and location. They also show with biological Matter accounted for, there is still an effect from location in addition to an interaction effect. Keeping this in mind, I decided to see how communities shape their microbiome between locations. High impacted sight and low impacted site. Panama (LowIMpact) and Colombia (HIMPACT). 
## These permutation compared all samples across coral compartments including outgroups, water and locations. 
uni_all_com_post <-pairwise.perm.manova(uni_all,all_out_df$political_area,nperm=1000)
uni_all_bio_post <-pairwise.perm.manova(uni_all,all_out_df$biological_matter,nperm=1000)

```
# *PERMANOVAs for Phylo_Complex and Robust
## 2. Comparisons across (Tissue, Skeleton ~ family*country_region)
```{r warning=FALSE}
## From initial pcoa and results, complex_robust and Sexual.system overlap and share similar impacts on data separation. In fact astrocoeniidae seems to drive these signficiant differences. Astrocoeniidae are both gonochores and spawners and considered complex corals. Mussidae and Merulinidae seem to drive the most difference in the skeleton microbiome community.
## Subset dataset to required variables 
#colombia <-subset_samples(ps.rarefied,political_area == "Colombia")
col_tis <-subset_samples(colombia, biological_matter == "Coral Tissue")
col_ske <-subset_samples(colombia, biological_matter == "Coral Skeleton")

## Create a dataframe for the adonis package below
col_df <- data.frame(sample_data(colombia))
col_tis_df <- data.frame(sample_data(col_tis))
col_ske_df <- data.frame(sample_data(col_ske))
#col_orbi_set <-subset_samples(ps.rarefied,family == "Merulinidae")

# Distance Matrix: All Colombian Samples
bray_col = phyloseq::distance(colombia, method="bray")
uni_col = phyloseq::distance(colombia, method="unifrac")
wuni_col = phyloseq::distance(colombia, method="wunifrac")

# Distance Matrix: Tissue compartments from Colombia
bray_col_ske = phyloseq::distance(col_ske, method="bray")
uni_col_ske = phyloseq::distance(col_ske, method="unifrac")
wuni_col_ske= phyloseq::distance(col_ske, method="wunifrac")

# Distance Matrix: Tissue compartments from Colombia Skeleton
bray_col_tis = phyloseq::distance(col_tis, method="bray")
uni_col_tis = phyloseq::distance(col_tis, method="unifrac")
wuni_col_tis = phyloseq::distance(col_tis, method="wunifrac")

# Oribcella distance matrix across colombian reefs high vs low impact
#bray_col_orb = phyloseq::distance(col_orbi_set, method="bray")
#uni_col_orb = phyloseq::distance(col_orbi_set, method="unifrac")
#wuni_col_orb = phyloseq::distance(col_orbi_set, method="wunifrac")

# Group distance matrix for loop
distance_methods_col <-c("bray_col_ske","uni_col_ske","wuni_col_ske ")
## For loop goes through making comparison across each distance matrix
# Must set the seed for reproducible results
set.seed(129)      

for (i in distance_methods_col){ 
  form <- as.formula(paste(i, "family*reef_name", sep="~"))
  print(form)
 adonis2(form, data=col_ske_df)->results_col
 print(results_col)
# capture.output(result, file = paste("C:/Users/Colin/Desktop/PSU Doctoral Program/GCMP_Working",i,"Colombian Coral Families Compartments #and Location",date(),".txt"))
}
col_tis_df$family
pan_tis_df$family

##col_ske_df$Sexual.system Skeleton samples were only collection from spawners found in colombia, 
##Skeleton was only sampled in Varadero so comparisons for skeleton will be limited to varadero
#col_tis_df$reef_name Tissue samples were collected from mutliple locations in Colombia, which shows a significant difference. 
## Complex_Robust has two levels for both tissue and skeleton.

## Pairwise perm Manova to test (dis)similarity between groups. 
#
bray_col_env_post <-pairwise.perm.manova(bray_col_ske, col_ske_df$family ,nperm=1000) 
uni_col_env_post <-pairwise.perm.manova(uni_col_ske,col_ske_df$family ,nperm=1000) 
wuni_col_env_post <-pairwise.perm.manova(wuni_col_ske,col_ske_df$family,nperm=1000)

# Accounting for presense absense and phylogeny, there are also differences. However those differences disapear when relative abundance comes into play
bray_col_ske_post <-pairwise.perm.manova(bray_col_ske, col_ske_df$complex_robust ,nperm=1000) 
uni_col_ske_post <-pairwise.perm.manova(uni_col_ske, col_ske_df$complex_robust ,nperm=1000) 
wuni_col_ske_post <-pairwise.perm.manova(wuni_col_ske, col_ske_df$complex_robust ,nperm=1000) 



```

# *PERMANOVAs for Environmental Comparison
## 3. Comparisons Pocillopora (Mucus ~ BiologicalMatter*country_region*host_genus )
```{r warning=FALSE}

# Group distance matrix for loop for Mucus
distance_methods_p_muc <-c("bray_pan_muc","wuni_pan_muc","uni_pan_muc")

## For loop goes through making comparison across each distance matrix
# Must set the seed for reproducible results
set.seed(129)      
#pan_tis_df$Mode.of.larval.development
for (i in distance_methods_p_muc){ 
  form <- as.formula(paste(i, "family*functional_group", sep="~"))
  print(form)
 adonis2(form, data=pan_muc_df)->results_pan_muc
 print(results_pan_muc)
# capture.output(result, file = paste("C:/Users/Colin/Desktop/PSU Doctoral Program/GCMP_Working",i,"Colombian Coral Families Compartments #and Location",date(),".txt"))
}

# Group distance matrix for loop: for Tissue
distance_methods_p_tis <-c("bray_pan_tis","wuni_pan_tis","uni_pan_tis")

## For loop goes through making comparison across each distance matrix
# Must set the seed for reproducible results
set.seed(129)      

for (i in distance_methods_p_tis){ 
  form <- as.formula(paste(i, "family*functional_group*Mode.of.larval.development", sep="~"))
  print(form)
 adonis2(form, data=pan_tis_df)->results_pan_tis
 print(results_pan_tis)
# capture.output(result, file = paste("C:/Users/Colin/Desktop/PSU Doctoral Program/GCMP_Working",i,"Colombian Coral Families Compartments #and Location",date(),".txt"))
}

# Group distance matrix for loop Skeleton
distance_methods_p_ske <-c("bray_pan_ske","wuni_pan_ske","uni_pan_ske")

## For loop goes through making comparison across each distance matrix
# Must set the seed for reproducible results
set.seed(129)      

for (i in distance_methods_p_ske){ 
  form <- as.formula(paste(i, "family*functional_group*Mode.of.larval.development", sep="~"))
  print(form)
 adonis2(form, data=pan_ske_df)->results_pan_ske
 print(results_pan_ske)
# capture.output(result, file = paste("C:/Users/Colin/Desktop/PSU Doctoral Program/GCMP_Working",i,"Colombian Coral Families Compartments #and Location",date(),".txt"))
}
##Pairwise perm Manova to test (dis)similarity between groups. 
# Mucus Comparisons. 
bray_pan_env_post <-pairwise.perm.manova(bray_pan_muc, pan_muc_df$family ,nperm=1000) 
uni_pan_env_post <-pairwise.perm.manova(uni_pan_muc,pan_muc_df$family ,nperm=1000) 
wuni_pan_env_post <-pairwise.perm.manova(wuni_pan_muc,pan_muc_df$family,nperm=1000)

# Tissue Accounting for presense absense and phylogeny, there are also differences. However those differences disapear when relative abundance comes into play
bray_pan_comp_post <-pairwise.perm.manova(bray_pan_tis, pan_tis_df$family ,nperm=1000) 
uni_pan_comp_post <-pairwise.perm.manova(uni_pan_tis, pan_tis_df$family ,nperm=10000) 
wuni_pan_comp_post <-pairwise.perm.manova(wuni_pan_tis, pan_tis_df$family ,nperm=1000) 

# Skeleton Accounting for presense absense and phylogeny, there are also differences. However those differences disapear when relative abundance comes into play
bray_pan_ske_post <-pairwise.perm.manova(bray_pan_ske, pan_ske_df$host_genus ,nperm=1000) 
uni_pan_ske_post <-pairwise.perm.manova(uni_pan_ske, pan_ske_df$host_genus ,nperm=1000) 
wuni_pan_ske_post <-pairwise.perm.manova(wuni_pan_ske, pan_ske_df$host_genus ,nperm=1000) 

## Both Sexual.systems and Mode.of.larval.development have multiple levels for tissue AND skeleton!
## Overall comparisons between tissue and skeleton with categorical variables show interesting results. First we see a siginficant differences in microbial community structure across tissue compartments when taking presense and absense into account. Moreover, taking both phylogeny and abundance metrics, we see that mucus, tissue and skeleton sampless taken from multiple coral families show significant differences. Showing that skeletonal abundance is siginficantly higher than tissue, and mucus samples and are more similar to sediment samples collected. Additionally phylogeny shows a small but significant effect is separation of the microbial communities across compartmennts. 
##Consistently, family show signifcantly different microbiomes, taking into account mode of sexual reproduction, functional group or mode of larval development. Overall, family showed to explain the most variation between families and that variation was siginficant. However, the remaining explanatory variables did not show significant differences between groups like sexual reproduction or mode of larval development. The closest that explain any significant varation after family was taken into account was functional group. When taking only phylygeny into account. ** What is interesting is, after comparing the tissue comparment, using manova perwise comparisons, showed no siginficant difference between family, genus or even species. I was surprised by this because I believed, these groups would be the most infuluenced by the host and therefore show significant differences between them. However, there was significnat differences in the skeleton of four major families. these differences and similarities also match phylogeny showing evidence of phylosymbiosis.  I'm now also thinking that the mucus across family may be where significant differences are and why potential functional groups are explaining more of the variation. # Note: 06/20/22 12:30p 

```

# Hclutering
```{r}
## Acropora muc Distance
bray_acro_muc = phyloseq::distance(acro_muc, method="bray")
uuni_acro_muc = phyloseq::distance(acro_muc, method="unifrac")
wuni_acro_muc = phyloseq::distance(acro_muc, method="Wunifrac")

## Acropora tis Distance
bray_acro_tis = phyloseq::distance(acro_tis, method="bray")
uuni_acro_tis = phyloseq::distance(acro_tis, method="unifrac")
wuni_acro_tis = phyloseq::distance(acro_tis, method="Wunifrac")

## Acropora ske Distance
bray_acro_ske = phyloseq::distance(acro_ske, method="bray")
uuni_acro_ske = phyloseq::distance(acro_ske, method="unifrac")
wuni_acro_ske = phyloseq::distance(acro_muc, method="Wunifrac")
acro_tis_df <-data.frame(sample_data(acro_tis))
acro_ske_df <-data.frame(sample_data(acro_ske))
# Clustering for: phylo_comp
hc_complex_muc <- hclust(uuni_acro_ske, method = "ward.D2")
# Get label names 
coral_family <- acro_ske_df$country_region
# Add tip labels
hc_complex_muc$labels <- coral_family
## Create the Dendrogram
dend <- as.dendrogram(hc_complex_muc)

#dend <- color_branches(dend, k=9)
#plot(dend)

##
dend %>% set("leaves_pch", 15) %>% 
  set("leaves_cex", 0.5) %>% # adjust the leave shape size
  set("labels_cex",1.0) %>% # adjust the leave label size
  color_branches(dend, k=4) %>%
  set("labels_col", k=4) %>%
  hang.dendrogram(hang_height = -0.002) %>% # hang level
  plot(main = "Caribbean Coral Microbiome Dendrogram \n Weighted Unifrac Distance with Mucus samples",
  horiz =  TRUE,  nodePar = list(cex = 0.01))

#### Clustering for: phylo_rob#####
hc_complex <- hclust(uni_ske, method = "ward.D2")
# Get label names 
coral_family <- ske_df$Huang_Roy_tree_name
# Add tip labels
hc_complex$labels <- coral_family
## Create the Dendrogram
dend <- as.dendrogram(hc_complex)

#dend <- color_branches(dend, k=9)
#plot(dend)

##
dend %>% set("leaves_pch", 15) %>% 
  set("leaves_cex", 0.5) %>% # adjust the leave shape size
  set("labels_cex",1.0) %>% # adjust the leave label size
  color_branches(dend, k=4) %>%
  set("labels_col", k=4) %>%
  hang.dendrogram(hang_height = -0.002) %>% # hang level
  plot(main = "Caribbean Coral Microbiome Dendrogram \n Weighted Unifrac Distance with Mucus samples",
  horiz =  TRUE,  nodePar = list(cex = 0.01))

#### Clustering for: comps vs robs #####

comp_rob_glom<- tax_glom(comps_vs_robs, taxrank = 'Family', NArm = TRUE)
bray_comp_rob <-tip_glom(comp_rob_glom)

hc_uni_comp_rob <- hclust(bray_comp_rob, method = "ward.D2")
# Get label names 
clade_family <- com_rob_df$country_region
# Add tip labels
hc_uni_comp_rob$labels <- clade_family
## Create the Dendrogram
dend3 <- as.dendrogram(hc_uni_comp_rob)
##
dend3 %>% set("leaves_pch", 15) %>% 
  set("leaves_cex", 0.5) %>% # adjust the leave shape size
  set("labels_cex",1.0) %>% # adjust the leave label size
  color_branches(dend3, k=4) %>%
  set("labels_col", k=4) %>%
  hang.dendrogram(hang_height = -0.002) %>% # hang level
  plot(main = "uuni Complex Robust family Cluster",
  horiz =  TRUE,  nodePar = list(cex = 0.01))

```

# Parking Lot
```{r}
## First, extract the top 100 ASVs
phylo_top50 <- prune_taxa(names(sort(taxa_sums(pan_muc),TRUE)[1:100]), pan_muc)

#phylo_noNA <- subset_taxa(ch_phylo, Phylum != "NA")
write.table(otu_table(phylo_top50), file = "otu_ofrank.txt", sep = "\t",
            row.names = TRUE)
otu_table <- read.table("otu_ofrank.txt", sep="\t", header=T, row.names=1)
otu_table.t <- t(otu_table)
#Calculate the percentage for each taxa
otus.perc <-otu_table.t/rowSums(otu_table.t)*100
## Clean out an NA from the og dataset
otus.perc_clean <- na.omit(otus.perc)

# create colour scale for heatmap
scaleyellowred <- colorRampPalette(c("lightyellow", "red"), space = "rgb")(100)

# make most basic possible heatmap
#quartz()
heatmap(as.matrix(otus.perc_clean), Rowv = NA, Colv = NA, col = scaleyellowred)
tis_matrix <- as.matrix(otus.perc_clean)
# identify the maximum abundance of each OTU for a single sample
maxab <- apply(otus.perc_clean, 2, max)

# for the purpose of display, remove OTUs with less than 1% max abundance
n1 <- names(which(maxab > 1))
data.prop.1 <- otus.perc_clean[, -which(colnames(otus.perc_clean) %in% n1)]

# Calculate inter-sample distance on overall dataset
data.dist <- vegdist(otus.perc_clean, method = "bray")
  


# Average linkage clustering
row.clus <- hclust(data.dist, "aver")

# Calculate which OTUs occur together more often for dual clustering
data.dist.g <- vegdist(t(otus.perc_clean), method = "bray")
col.clus <- hclust(data.dist.g, "aver")

# Plot only abundant OTUs, but clustering from whole dataset
#quartz()
heatmap(as.matrix(otus.perc_clean), Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus), 
        col = scaleyellowred, margins = c(10, 3))
## Plot the heatmap
plot_heatmap(phylo_top50, "Phylum")
plot_heatmap(phylo_top50, "NMDS", "bray", "SampleType", "Family")
# Let's create a smaller dataset based on a specific phylum. Here, I am choosing Proteobacteria.
##phylo_prot <- subset_taxa(phylo_nolow, Phylum == "Proteobacteria")


```
## Parking LotAlpha Diversity BoxPlots Species
```{r}
###Once we have our sub_samples rarefied we can start with some diversity analysis  and visualization
### Alpha Diversity ** you can make multiple diversity box plots by switch x for sample variables
#1
astr_diversity_compartment <- plot_richness(astr, x="biological_matter", title = "Astrocoeniidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))
#2
meru_diversity_compartment <- plot_richness(complex, x="family", color = "family", title = "Merulinidae Diversity Across compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))
#3
mon_diversity_compartment <- plot_richness(mon, x="biological_matter", title = "Montastreidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))
#4
muss_diversity_compartment <- plot_richness(muss, x="tissue_compartment", title = "Mussidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

#5
sidsp_diversity_compartment <- plot_richness(sidsp, x="biological_matter", title = "Siderastreidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

#6
porpor_diversity_compartment <- plot_richness(porpor, x="biological_matter", title = "Poritidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))
#7
mea_diversity_compartment <- plot_richness(mea, x="biological_matter", title = "Meandrinidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

#8
agri_diversity_compartment <- plot_richness(agri, x="biological_matter", title = "Agariciidae Diversity Across compartment", color = "tissue_compartment", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

#9
acrop_diversity_compartment <- plot_richness(acrop, x="host_name", title = "Acroporidae Diversity Across Species Tissue", color = "host_name", measures = c("Shannon","Observed"),)+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90))

```
## Script Parkinglot
```{r}
## Parkinglot

## Compartments
#mucus <-subset_samples(main_coral,BiologicalMatter == "Coral Mucus")
#Tissue Subset
#tissue <-subset_samples(main_coral,BiologicalMatter == "Coral Tissue")
#Skeleton Subset
#skeleton <-subset_samples(main_coral,BiologicalMatter == "Coral Skeleton")
##Distance matrix for complex_Family_Compartment
bray_acro_muc = phyloseq::distance(acro_muc, method="bray")
bray_acro_tis = phyloseq::distance(acro_tis, method="bray")
bray_acro_ske = phyloseq::distance(acro_ske, method="bray")
uni_acro_tis = phyloseq::distance(acro_tis, method="unifrac")

##Distance matrix for complex_Family_Compartment
bray_agra_muc = phyloseq::distance(agra_muc, method="bray")
bray_agra_tis = phyloseq::distance(agra_tis, method="bray")
bray_agra_ske = phyloseq::distance(agra_ske, method="bray")
uni_agra_tis = phyloseq::distance(agra_tis, method="unifrac")

##Distance matrix for Robust_Family_Compartment
bray_meru_muc = phyloseq::distance(meru_muc, method="bray")
bray_meru_tis = phyloseq::distance(meru_tis, method="bray")
bray_meru_ske = phyloseq::distance(meru_ske, method="bray")
uni_meru_tis = phyloseq::distance(meru_tis, method="unifrac")

##Distance matrix for Robust_Family_Compartment
bray_poci_muc = phyloseq::distance(poci_muc, method="bray")
bray_poci_tis = phyloseq::distance(poci_tis, method="bray")
bray_poci_ske = phyloseq::distance(poci_ske, method="bray")
uni_poci_tis = phyloseq::distance(poci_tis, method="unifrac")


#robust_group_mucus
meru_muc <-subset_samples(mucus, family == "Merulinidae")
poci_muc <-subset_samples(mucus,family == "Pocilloporidae")
dipl_muc <-subset_samples(mucus,family == "Diploastreidae")
#robust_group_tissue
meru_tis <-subset_samples(tissue, family == "Merulinidae")
poci_tis <-subset_samples(tissue,family == "Pocilloporidae")
dipl_tis <-subset_samples(tissue,family == "Diploastreidae")
#robust_group_skeleton
meru_ske <-subset_samples(skeleton, family == "Merulinidae")
poci_ske <-subset_samples(skeleton,family == "Pocilloporidae")
dipl_ske <-subset_samples(skeleton,family == "Diploastreidae")

#complex_group_mucus
acro_muc <-subset_samples(mucus, family == "Acroporidae")
por_muc <-subset_samples(mucus, family == "Poritidae")
agra_muc <-subset_samples(mucus, family == "Agariciidae")
#dendro <-subset_samples(main_coral,family == "Dendrophylliidae")

#complex_group_tissue
acro_tis <-subset_samples(tissue, family == "Acroporidae")
por_tis <-subset_samples(tissue, family == "Poritidae")
agra_tis <-subset_samples(tissue, family == "Agariciidae")
#dendro <-subset_samples(main_coral,family == "Dendrophylliidae")

#complex_group_skeleton
acro_ske <-subset_samples(skeleton, family == "Acroporidae")
por_ske <-subset_samples(skeleton, family == "Poritidae")
agra_ske <-subset_samples(skeleton, family == "Agariciidae")
#dendro <-subset_samples(main_coral,family == "Dendrophylliidae")

# Clade
rob_df <- data.frame(sample_data(robust))
com_df <- data.frame(sample_data(complex))

#Robust Family 
meru_df <- data.frame(sample_data(meru))
lobo_df <- data.frame(sample_data(lobo))
poci_df <- data.frame(sample_data(poci))
dipl <- data.frame(sample_data(dipl))
## Complex Family
acro_df <- data.frame(sample_data(acro))
por_df <- data.frame(sample_data(por))
agra_df <- data.frame(sample_data(agra))
dendro_df <- data.frame(sample_data(dendro))


mon <-subset_samples(main_coral, family == "Montastreidae") ### Only tissue and samples from panama contain one sample set
astr <-subset_samples(phylo_nclass, family == "Astrocoeniidae")
sidsp <-subset_samples(phylo_nclass, family == "Siderastreidae")
mea <-subset_samples(phylo_nclass, family == "Meandrinidae") 
#tis_df$Mode.of.larval.development

#carib <- subset_samples(main_coral, ocean == "Atlantic")
#east_panama <-subset_samples(main_coral, expedition_number == "E2")
#west_panama <-subset_samples(main_coral, expedition_number == "E11")
#colombia <-subset_samples(main_coral, political_area == "Colombia")
#australia <-subset_samples(main_coral, political_area == "Australia")
#hawaii <-subset_samples(main_coral, political_area == "Hawaii, United States")
#lobo <-subset_samples(main_coral,family == "Lobophylliidae")
#muss <-subset_samples(main_coral, family == "Mussidae")

```
## Phyloseq Metadata Query
```{r}
## To obtain metadata from subset samples 
 meru_metadata <-(sample_data(meru))
view(meru_metadata)

# Total number of sequences
print("total")

sum(sample_sums(ch_phylo))
ch_phylo
# Min number of sequences in a sample
print("min")
min(sample_sums(ch_phylo))

# Mean number of sequences
print("mean")
mean(sample_sums(ch_phylo))

# Median number of sequences
print("median")
median(sample_sums(ch_phylo))

# Max number of sequences
print("max")
max(sample_sums(ch_phylo))

# Total number of ASVs
ntaxa(ch_phylo)

```
